pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/unsafe/MmapMemory.java

```diff
@@ -0,0 +1,348 @@
+            }
+
+            return new MapSection(address, unmapFun);
+          } catch (InvocationTargetException | IllegalAccessException e) {
+            throw new RuntimeException(e);
+          }
+        };
+      }
+    }
+  }
+
+  /**
+   * A {@link MapFun} that actually delegates into a map0 native method included in pre 19 Java releases.
+   *
+   * Unlike normal map methods, map0 actually has some low level requirements. For example, the offset must be page
+   * aligned.
+   */
+  interface Map0Fun extends MapFun {
+
+    /**
+     * @param offset It has to be a positive value that is page aligned.
+     */
+    MapSection map0(FileChannel fc, boolean readOnly, long offset, long size)
+        throws InvocationTargetException, IllegalAccessException, IOException;
+
+    default MapSection map(File file, boolean readOnly, long offset, long size) throws IOException {
+      String mode = readOnly ? "r" : "rw";
+      try (RandomAccessFile raf = new RandomAccessFile(file, mode); FileChannel fc = raf.getChannel()) {
+        if (size == 0) {
+          return MapSection.EMPTY;
+        }
+
+        long allocationGranule = Unsafer.UNSAFE.pageSize();
+        int pagePosition = (int) (offset % allocationGranule);
+
+        // Compute mmap address
+        if (!fc.isOpen()) {
+          throw new IOException("closed " + file.getPath());
+        }
+
+        long fileSize = fc.size();
+        if (fileSize < offset + size) {
+          // If file size is smaller than the specified size, extend the file size
+          raf.seek(offset + size - 1);
+          raf.write(0);
+          //logger.trace(s"extend file size to ${fc.size}")
+        }
+        long mapPosition = offset - pagePosition;
+        long mapSize = size + pagePosition;
+        // A workaround for the error when calling fc.map(MapMode.READ_WRITE, offset, size) with size more than 2GB
+
+        MapSection map0Section = map0(fc, readOnly, mapPosition, mapSize);
+        return new MapSection(map0Section.getAddress() + pagePosition, map0Section.getUnmapFun());
+      } catch (InvocationTargetException | IllegalAccessException e) {
+        throw new RuntimeException("Cannot map file " + file + " from address " + offset + " with size " + size, e);
+      }
+    }
+
+    static BiConsumer<Long, Long> tryFindUnmapper()
+        throws NoSuchMethodException, ClassNotFoundException {
+      Class<?> fileChannelImpl = MmapMemory.class.getClassLoader().loadClass("sun.nio.ch.FileChannelImpl");
+      Method unmapMethod = fileChannelImpl.getDeclaredMethod("unmap0", long.class, long.class);
+      unmapMethod.setAccessible(true);
+      return (address, size) -> {
+        try {
+          unmapMethod.invoke(null, address, size);
+        } catch (IllegalAccessException | InvocationTargetException e) {
+          throw new RuntimeException(e);
+        }
+      };
+    }
+
+    class ChronicleCore implements Finder<Map0Fun> {
```

Comment:
previous experiments / attempts by you had concluded that Chronicle Bytes has several blocking bugs and was not suitable for use by us. What changed since then that we are back to using it here and also preferring it to be the default option?
I thought that option 4 was the obvious choice, but seems like I'm missing something

Reply:
Chronicle has several libraries. For this use case we care about 3 of them:
- Chronicle Posix is a JNA wrapper that let us call the POSIX API.
- Chronicle Core is a utility library used in by their other libraries. It depends on POSIX. It offers some OS and JVM utilities that may be useful. For example, does the same reflection tick I had to do to find the mmap methods.
- Chronicle Bytes is a buffer library built on top of Chronicle Core. In the Chronicle ecosystem it plays the role of our PinotDataBuffer.

My first attempt was to use Chronicle Bytes as the backend of my PinotByteBuffer library and it resulted in several strange semantic bugs. Chornicle Core is quite simpler and provides access to low level stuff. That is why I'm proposing to use it. Chronicle Posix is just a wrapper on top of the POSIX api, so it is a very simple library and difficult to make mistakes there.